### PR TITLE
[GH-257] Display labels and assignee on new issues and PRs

### PR DESCRIPTION
**Pull request by Szymongib**
#### Summary
Display labels and assignee when creating new Issue or PR.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-github/issues/257

### CODE REVIEW

**Comment by Szymongib** [comment]
Hi @mickmister, 
I have a couple of follow up questions to finalize PR:
- Do you have any more suggestions regarding how labels and assignee should be displayed?
Currently, it looks like that:
image
- Do you think those should be displayed in case of any other event as well?

**Comment by mickmister** [comment]
@Szymongib A quick pass through the code looks good :+1:

> Do you have any more suggestions regarding how labels and assignee should be displayed?

Do you mind posting some screenshots of the entire post, one with lots of summary/description text, and with a small amount?

> Do you think those should be displayed in case of any other event as well?

I think the additions would make sense on new comment posts as well, that are sent to a subscribed channel. As opposed to a DM notification from a comment.
@asaadmahmood What do you think?

What do you think about showing the other fields that can be set on an issue/PR? Such as Projects, Milestones, and Linked issues/PRs?

**Comment by mickmister** [comment]
I'm thinking this has a lot of overlap with the plugin tooltip display, but this PR's functionality would show on mobile. Also, a post by the GitHub bot is more significant than a single link to an issue/PR, which is the scenario for the tooltip.

We may want to keep it simple with just labels and assignees. @Szymongib I'm wondering how it looks in the mobile app with more fields showing

**Comment by asaadmahmood** [comment]
@mickmister @Szymongib Let's get a full screenshot of the channel/center channel to see how it looks.
Also @mickmister please help me explain this a bit more. I'm assuming this information will be displayed as a message to the channel (if subscribed) or DM.

What would be the case where this appears as a comment?

**Comment by mickmister** [comment]
@asaadmahmood 
> Also @mickmister please help me explain this a bit more. I'm assuming this information will be displayed as a message to the channel (if subscribed) or DM.

Right now, it is showing for new PRs and new issues. If a channel is subscribed to pull request or issue creation, there will be posts in that channel and the labels will show up with this PRs functionality.

> What would be the case where this appears as a comment?

It's just a thought of having them show for comments as well. Channels can subscribe to GitHub comment events, and the labels could be shown with those posts. DM notifications could also have them, but I think that will be bloated for the GitHub DM feed.

**Comment by asaadmahmood** [comment]
@mickmister I think it makes sense for issue creation and Pull requests, but I think for every comment, it may not be the best for us, as we may get a duplication on that most of the time.

I feel like we can have a separate event (if we don't have that right now) for label updates or assignee updates that can be used if people want to record those activities as well.

**Comment by mickmister** [comment]
@asaadmahmood Currently there is no event for label or assignee changes. Oftentimes, labels are added by the reporter  after issue creation. So having an event for that would be nice. It may be within scope of task since it is directly related to this use case.

**Comment by Szymongib** [comment]
@mickmister I think showing additional information like Project or Milestone could be added later as an enhancement to first see feedback from adding labels and assignee.

Here are some screenshots:

long issue


short issue


PR:

PR


and from the mobile app (IOS):

long issue


short issue


PR

**Comment by mickmister** [comment]
@asaadmahmood Here are some screenshots ^

@Szymongib Placement of the new items looks good to me

**Comment by mickmister** [comment]
@Szymongib Maybe place commas between each label? Just a thought, not sure if it is right to do so

**Comment by Szymongib** [comment]
> Maybe place commas between each label? Just a thought, not sure if it is right to do so

Sure @mickmister, I think that makes sense, especially for mobile version.

Here are some screenshots for comparison:

image

mobile

**Comment by Szymongib** [comment]
@mickmister, @asaadmahmood
Regarding the events for label updated I am not sure how it would suppose to work as I believe the GitHub would emit an event for every label added so if the 3 labels are added to the issue it would result in 3 posted messages. I feel like it could get really spammy quite quickly.
What are your thoughts about it?

**Comment by asaadmahmood** [comment]
@Szymongib If they emit 3 events, then yes, we can post 3 events. But I think when a person is setting multiple labels together, it may just emit a single event.
So just something to look at.

The designs high level look good. On mobile, I think we have more spacing between the labels and the new issue text, but if not possible, its fine as is.

**Comment by Szymongib** [comment]
@asaadmahmood You are right if multiple labels are assigned at once only one event is emitted.

Sorry, I do not have enough knowledge about the plugin, so correct me if I am wrong. As I understand the functionality would post a message to the subscribed channel when the issue/PR is labeled or assigned?

If that is the case I am not totally convinced if that is a good move, as it may result in a lot of messages and I am not sure if those would bring much value. Maybe making it somehow optional would make more sense?

**Comment by asaadmahmood** [comment]
Most definitely, I'm fine if we can make that optional in the subscription.

**Comment by Szymongib** [comment]
I have found that there already is the functionality of posting messages when the issue is labeled. By subscribing to feature label:"[LABEL_NAME]" the message is posted to the channel when the issue/PR is labeled with `[LABEL_NAME]`.
I feel like the new functionality would kind of conflict/duplicate already existing feature.

I have reached out to @mickmister to align how to proceed with this PR.

File: server/plugin/template.go [hunk 1]
```diff
@@ -149,10 +169,24 @@ func init() {
...
 
+	template.Must(masterTemplate.New("labels").Funcs(funcMap).Parse(`
+{{- if .Labels }}
+Labels: {{range $i, $el := .Labels -}}` + "{{- if $i}}, {{end}}[`{{ $el.Name }}`]({{ $.RepositoryURL }}/labels/{{ $el.Name | pathEscape }})" + `{{end -}}
+{{ end -}}
+`))
+
+	template.Must(masterTemplate.New("assignee").Funcs(funcMap).Parse(`
+{{- if .GetAssignee }}
+Assignee: [{{ .GetAssignee.GetLogin }}]({{ .GetAssignee.GetHTMLURL }})
```
Comment (mickmister) [inline server/plugin/template.go:180]: It's possible that an issue/PR has multiple assignees. Is that taken into account here?

Comment (Szymongib) [reply server/plugin/template.go]: Fair point. As I understand `GetAssignee` returns only a single one. I have modified it so that all assignees are

**Comment by Szymongib** [comment]
We agreed, that the functionality with labels changes will be implemented as part of a different issue https://github.com/mattermost/mattermost-plugin-github/issues/340

File: server/plugin/template.go [hunk 2]
```diff
@@ -149,10 +169,24 @@ func init() {
...
 
+	template.Must(masterTemplate.New("labels").Funcs(funcMap).Parse(`
+{{- if .Labels }}
+Labels: {{range $i, $el := .Labels -}}` + "{{- if $i}}, {{end}}[`{{ $el.Name }}`]({{ $.RepositoryURL }}/labels/{{ $el.Name | pathEscape }})" + `{{end -}}
+{{ end -}}
+`))
+
+	template.Must(masterTemplate.New("assignee").Funcs(funcMap).Parse(`
+{{- if .Assignees }}
+Assignees: {{range $i, $el := .Assignees -}} {{- if $i}}, {{end}}[{{ $el.GetLogin }}]({{ $el.GetHTMLURL }}){{end -}}
```
Comment (hanzei) [inline server/plugin/template.go:180]: If the user has a connected GitHub account, should this link to the Mattermost account instead?

[the remark this answers is not in the document]
Comment (Szymongib) [reply server/plugin/template.go:180]: Sure, if the index is 0 - meaning single assignee, this evaluates to false and the coma is not added in front of the assignee.

Comment (Szymongib) [reply server/plugin/template.go:180]: Right, I see that you are using such a convention in other places where the user is displayed. I think it makes sense to change it so it is consistent unless anyone has some objections?

[the remark this answers is not in the document]
Comment (Szymongib) [reply server/plugin/template.go:180]: So this prevents adding a coma when there is for example only one assignee.

Comment (hanzei) [reply server/plugin/template.go]: Makes sense :+1:

**Pull request merged**